src/emulator/storage/files.ts

```diff
@@ -335,23 +335,29 @@ export class StorageLayer {
   }
 
   /**
-   * Stores the uploaded file with generated metadata and triggers Object Finalize Cloud Functions.
+   * Stores the uploaded file after finalizing metadata and triggers Object Finalize Cloud Functions.
    * @param upload The upload to finalize.
    * @returns The stored file.
    */
   public finalizeUpload(upload: ResumableUpload): StoredFile {
     upload.status = UploadStatus.FINISHED;
 
-    const metadata = this.createMetadata(upload);
+    const storedMetadata = this.createMetadata(upload);
+    const incomingMetadata = upload.metadata;
+    storedMetadata.update(incomingMetadata);
+    if (storedMetadata.downloadTokens.length === 0) {
```

Comment:
`finalizeUpload` is also used by `gcloud.ts` which should not generate a new download token on upload. I would move this back to `firebase.ts`

Reply:
Good catch, I'll move the token generation back. I think the `storedMetadata.update(incomingMetadata);` should be fine to keep as `gcloud.ts` and `firebase.ts` both take the metadata response and return it

Reply:
Scratch what I said, didn't end up doing what I suggested it after looking at the rebased code again